feature: Add presence markers to PM compose.

plugyawn opened the pull request:
<!-- Please see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->

**What does this PR do?**  <!-- Overall description goes here -->

Adds presence markers to PM compose.

Previously the only way to know if the person on the other side was online was to check through the info window or through the sidebar, which are big hassles. In the future, for non-group PMs, we could add a good last-seen implementation that doesn't take up unnecessary space. 


<!-- If fixing a filed bug or new feature, add 'Fixes #<issue>' or 'Partial fix for #<issue>' -->

Partial fix for #1155.

<!-- Add a link to a discussion on chat.zulip.org, if relevant -->

**Tested?** <!-- Fine to leave some of these unchecked if this is a draft/work-in-progress -->
- [x] Manually
- [ ] Existing tests (adapted, if necessary)
- [ ] New tests added (for any new behavior)
- [ ] Passed linting & tests (each commit)
<!-- Code must pass CI (GitHub Actions) before merging - look for the green tick! -->

<!-- See https://github.com/zulip/zulip-terminal#commit-style -->
**Commit flow** <!-- if more than one commit; add/delete/fill-in as appropriate -->
<!-- For example:
- maybe multiple commits doing similar things
-->
- Add presence markers to PM compose


**Notes & Questions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- For example:
- this doesn't include feature X (yet?)
- unsure about Y
- should this do Z?
-->
- markers don't have colours yet. 

**Visual changes** <!-- if any; add/delete/fill-in with screenshot/diagram as appropriate -->
<img width="369" alt="Screenshot 2022-04-07 at 7 24 27 PM" src="https://user-images.githubusercontent.com/76529011/162218059-f64da147-da17-4d98-96fc-3bcbf169ff9c.png">

plugyawn committed:
feature: Add presence markers to PM compose.

Fixes part of #1155.

PMs didn't have a way to know whether the user was active without presence markers from the right column view, which is a hassle. This allows them to see it directly.

Need to add colour to the markers.

In the future, maybe add some form of last seen text that's shorter than just the timestamp ("last seen 3 hours ago") instead of ("last seen on Fri 3 Aug 03-02-2002").

prah23 commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -201,16 +201,14 @@ def private_box_view(
             recipient_info = ", ".join(
                 [
-                    f"{self.model.user_dict[email]['full_name']} <{email}>"
+                    f"""{STATE_ICON[self.model.user_dict.get(email, None).get("status","inactive") if self.model.user_dict.get(email,None) else "inactive"]} {self.model.user_dict[email]['full_name']} <{email}>"""
```
The text being added here is validated before any action is performed, such as sending a message, saving a message as a draft, or even cycling out of the recipient header. Only adding an icon here will not bypass the validation check and hence, prevent the user from performing any of the aforementioned actions.

plugyawn replied:
@prah23 thanks for pointing that out! Sorry, I thought I fixed this already, but I just realised I forgot to push it. I'll do so as soon as I get access to my computer again. My bad.

prah23 commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -201,16 +201,14 @@ def private_box_view(
             self.recipient_emails = []
             recipient_info = ""
-
-        self.send_next_typing_update = datetime.now()
```
Unrelated?

plugyawn replied:
Unrelated. Forgot to add a commit to fix the errors that sprung up.

prah23 reviewed (changes requested):
Thanks for the PR, @plugyawn.

Left a couple of comments inline, but more broadly, I'd suggest testing the proposed feature as a user thoroughly, so that any unintended behavior is detected and dealt with. The change proposed here currently prevents fundamental user abilities such as sending a private message.

plugyawn committed:
error fixes: message sending wasn't working before.

plugyawn commented:
@prah23 Sorry, could you check again now?

plugyawn committed:
reformat and error fixes: reformatted with black and fixed linting error.

prah23 commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -201,17 +201,26 @@ def private_box_view(
             recipient_info = ", ".join(
                 [
-                    f"{self.model.user_dict[email]['full_name']} <{email}>"
+                    f"""{self.model.user_dict[email]['full_name']} <{email}>"""
+                    for email in self.recipient_emails
+                ]
```
There's no change here; it can be removed from the commit.

plugyawn replied:
Oops -- will do.

prah23 reviewed (changes requested):
Thanks for the update.

The current iteration would make it hard to understand which status marker belongs to which user when there's a group PM of > 4-5 people, since all the markers are shown at the beginning and the names + emails come afterwards.  There is no update in the number of status markers either when a new recipient is added; if I add a third recipient to a group PM with 2 current recipients, there are still only 2 markers.

I would suggest discussing the design before updating the PR, so that you have more feedback to work with beforehand.

plugyawn commented:
<img width="603" alt="Screenshot 2022-04-08 at 6 27 07 PM" src="https://user-images.githubusercontent.com/76529011/162440350-c6c9a827-25f5-4c85-9e21-3979a330a3a0.png">

Fixed the issue with Group PMs!

plugyawn committed:
boxes: Fixed issue with repeating state markers.

plugyawn committed:
boxes.py: fixed bugs.

Was encountering a bug where
- sending message would cause the state_markers to disappear.
- message won't be sent because the recipient_name had the state_marker appended to it.

Fixed both by making relevant changes to _tidy_valid_recipients_notify_invalid_ones() which gets called every time composer hits send.

plugyawn commented:
Needs some changes in the tests.

plugyawn committed:
tests: Updated tests for boxes.py

Added presence markers to tests.

plugyawn committed:
reformatting: used black to reformat the code.
